### PR TITLE
Multiple Components

### PR DESCRIPTION
**Pull request by Delerme**
Refactors storage and various other things to support multiple filter components

### CODE REVIEW

File: src/ui/components/componentmanager.js
Comment (b1lly) [inline src/ui/components/componentmanager.js]: Two nits:

1. Probably should be a function instead of a property
2. Yext style for naming variables/functions is typically:  [verb][noun]

```
if (componentClass.allowDuplicates()) {

}
```

I have mixed feelings about "public properties" in Javascript. They should be used sparingly.

File: src/ui/components/componentmanager.js
Comment (b1lly) [inline src/ui/components/componentmanager.js]: The name here doesn't describe what it's checking. As someone reading this code, I need to dig into the implementation to see whats going one. As a general coding hygiene, you want to try to make your code a bit more prosaic so that other people reading your code don't need to dive into implementation details. 

Additionally, I think this methods contract/behavior seems unintuitive. The caller of the method should decide whether or not to throw an exception, and the function should just return true or false.

Probably better to do something like this:

```
if (this._isDuplicateName(componentType, opts.name)) {
  // throw error here
}
```

And your implementation:

```
isDuplicateName(componentType, name) {
  if (//duplicate) { 
    return true;
  }
  return false;
}
```

Then you can test it easily, its name and what it does is obvious. The caller decides how to handle the response, and the contract is clear and concise from the top level without having to read the details. The method name describes exactly what its checking for, and the result of the function is expected.

File: src/ui/components/componentmanager.js
Comment (b1lly) [inline src/ui/components/componentmanager.js:135]: Did you test SearchComponent to make sure its working properly? I believe its using this method.

Side question: What if there are two components of the same type? Does this interface break? We might want to consider that use case. I don't think it would be something that happens often but if it is, this certainly wouldn't work properly here.

File: src/ui/components/search/autocompletecomponent.js
Comment (b1lly) [inline src/ui/components/search/autocompletecomponent.js]: Maybe we should take in a configuration object now as the interface. That's a lot of arguments.

I'd also like to add tests for these methods if possible.

File: src/core/core.js
Comment (b1lly) [inline src/core/core.js:81]: I mentioned this in an earlier comment, not sure if there was a response. It might have gotten lost since the comment wasn't on a PR.

But, do we want to make a bulk insert interface?

e.g.
```
this.storage.insert({
  [StorageKeys.NAVIGATION]: data[StorageKeys.NAVIGATION],
  [StorageKeys.DIRECT_ANSWER]: data[StorageKeys.DIRECT_ANSWER],
  ....
})
```

And also support a single insert

```
this.storage.insert(StorageKeys.NAVIGATION, data[StorageKeys.NAVIGATION])
```

I think it's more convenient, and you also get performance gains since object resolution in Javascript is generally slow.

File: src/core/storage/storage.js
Comment (b1lly) [inline src/core/storage/storage.js:18]: If someone doesn't provide a key, or data, how do you want to handle that here?

File: tests/core/storage/storage.js
Comment (b1lly) [inline tests/core/storage/storage.js]: Isn't this the same test as stores data by the provided key?

File: tests/core/storage/storage.js
Comment (b1lly) [inline tests/core/storage/storage.js]: Just to confirm my understanding, whats the purpose of the key in data? It doesn't have one right? its just data to be tested right? The fact that the data is key: [keyVal] doesn't matter, right? That was a little confusing.

File: tests/core/storage/storage.js
Comment (b1lly) [inline tests/core/storage/storage.js]: do we want to update these names from module to something different?

File: README.md
Comment (b1lly) [inline README.md]: unique

File: src/core/core.js
Comment (b1lly) [inline src/core/core.js]: probably time to make this a configuration object. I think 4 arguments is definitely pushing the fold. barKey is optional, right?

File: src/core/core.js
Comment (b1lly) [inline src/core/core.js]: same as above.

File: src/core/core.js
Comment (b1lly) [reply src/core/core.js]: You also might want to JSDoc the arguments and purpose. I'm not sure its clear to me what autoCompleteName is from looking at the signature.

File: src/core/core.js
Comment (b1lly) [inline src/core/core.js]: I actually don't think the core should care about "picker" as a thing. The picker is an implementation detail of the component ui library and not the core.

In other words, i'd do something more like:

```
applyFilter(namespace, verticalKey, filter)
```

Also see my structural message.

File: src/ui/components/componentmanager.js
Comment (b1lly) [inline src/ui/components/componentmanager.js]: See above comment.

File: src/ui/components/search/autocompletecomponent.js
Comment (b1lly) [inline src/ui/components/search/autocompletecomponent.js:259]: This looks like the same exact logic as above in the DOM focus, can we encapsulate this somewhere and just use it in both places?

File: src/ui/components/search/filterpickercomponent.js
Comment (b1lly) [inline src/ui/components/search/filterpickercomponent.js]: OK, I think we just introduced some dependencies to the search interface, and some wonky side effects that I don't think we actually want here.

The search interface should be simple. 

1. It takes input (query and filters),
2. issues a network request. 

Having it magically insert into storage and pull data out of storage to build the request probably isn't what we want to do here. We want to do that stuff before it gets here and the provide it as input.

File: src/ui/components/search/filterpickercomponent.js
Comment (b1lly) [inline src/ui/components/search/filterpickercomponent.js:185]: How come the query is name spaced? Shouldn't there only ever be one query?

**Review by b1lly — COMMENTED** [review]
Nice job dude.

A few minor style things and there's one main structural thing that I think needs to be addressed.

The structural thing is how the search interface and implementation actually works. We can talk offline if there's any confusion but I think what you have here is confusing and error prone. I don't think the search flow should have a dependency on storage at all. I think the information it needs to issue a network request should be provided to it as input. Doing some magic inside the search flow to set and get the state of filters in order to build the request doesn't seem like the right thing to do.  It's unnecessary complexity and can definitely have some weird side effects and race conditions in the future.

I think all that complexity of updating state and retrieving state of filters should be handled outside of the search path (before it), and instead should be provided to it.

File: src/ui/components/componentmanager.js
Comment (Delerme) [reply src/ui/components/componentmanager.js]: done

File: src/ui/components/componentmanager.js
Comment (Delerme) [reply src/ui/components/componentmanager.js]: done

File: src/ui/components/componentmanager.js
Comment (Delerme) [reply src/ui/components/componentmanager.js:135]: yea that's why I kept this function at all. Does not work with two components of the same type, but neither did the original implementation. I notice you had a TODO for removing the need for this function, so I think we should close out that TODO and delete this function entirely

File: src/ui/components/search/autocompletecomponent.js
Comment (Delerme) [reply src/ui/components/search/autocompletecomponent.js]: I split out each autocomplete to make it easier to see what's going on (the component now contains the logic for deciding which core method to call). As discussed a config object doesn't help a lot in this case because all of the arguments are required, so the config object makes it harder to reason about the function and see exactly what it needs. Barkey will be deprecated soon which will bring these methods back down to three args as well

File: src/core/core.js
Comment (Delerme) [reply src/core/core.js:81]: I considered it but I prefer the insertions to be explicit. It's much easier to see what data is getting inserted under which key, instead of hunting down all the keys that have been put on some object. It also opens up room for bugs like accidentally modifying the object and putting data we don't want stored, etc. I can investigate some nicer ways of handling this if it comes up again

File: src/core/storage/storage.js
Comment (Delerme) [reply src/core/storage/storage.js:18]: Added preconditions with error throwing

File: tests/core/storage/storage.js
Comment (Delerme) [reply tests/core/storage/storage.js]: yea I guess so, removed

File: tests/core/storage/storage.js
Comment (Delerme) [reply tests/core/storage/storage.js]: Yea I just copied over the tests you had. For clarity though I changed all the test data to avoid this confusion in the future

File: tests/core/storage/storage.js
Comment (Delerme) [reply tests/core/storage/storage.js]: yea probably, done

File: README.md
Comment (Delerme) [reply README.md]: done

File: src/core/core.js
Comment (Delerme) [reply src/core/core.js]: see above, added better docs

File: src/core/core.js
Comment (Delerme) [reply src/core/core.js]: done

File: src/core/core.js
Comment (Delerme) [reply src/core/core.js]: done

File: src/ui/components/componentmanager.js
Comment (Delerme) [reply src/ui/components/componentmanager.js]: done

File: src/ui/components/search/autocompletecomponent.js
Comment (Delerme) [reply src/ui/components/search/autocompletecomponent.js:259]: done

File: src/ui/components/search/filterpickercomponent.js
Comment (Delerme) [reply src/ui/components/search/filterpickercomponent.js]: done

File: src/ui/components/search/filterpickercomponent.js
Comment (Delerme) [reply src/ui/components/search/filterpickercomponent.js:185]: nope, since there can be several filter pickers there can be several text inputs, and each one has a separate query. I.e. box1 has 'software engineer' and box2 has 'tysons, virginia'. We want each one stored individually so if I link to the page each box gets populated with the correct value

File: src/ui/components/search/autocompletecomponent.js
Comment (Delerme) [reply src/ui/components/search/autocompletecomponent.js]: Also in terms of tests for this component, I have to figure out
- loading handlebar templates into jest from source files (you have a todo for this already)
- abstracting out the test environment so we don't need as much setup, and can render our components like normal (will probably use enzyme for this)
- ensuring the full component works with mocked out interfaces

So I'll be working on that now but won't include it in this commit, as there are a lot of unknowns and there's no need for that to block this

Comment (b1lly) [reply src/ui/components/search/autocompletecomponent.js]: Testing the component would be ideal, but I agree theres a ton of complexity there. I think there's a ton of logic the component depends on that we can start testing already without having to figure out a lot of those things. For instance, we don't have to necessarily test the component in order to add tests to the core methods (e.g. autoCompleteFilter). I think those would be a good starting point.

File: src/ui/components/componentmanager.js
Comment (b1lly) [reply src/ui/components/componentmanager.js]: FWIW I thought having a function that encapsulated the meaning of the expression was a cool thing to have. I was just saying the name could've been improved, and the contract in/out could be adjusted.

Either way thought works.

Comment (Delerme) [reply src/ui/components/componentmanager.js]: ack

File: src/ui/components/search/autocompletecomponent.js
Comment (Delerme) [reply src/ui/components/search/autocompletecomponent.js]: yea core just calls out to storage which I added tests for, and autocomplete which I will add tests to when I figure out a good way to mock out the api request stuff

File: src/core/core.js
Comment (b1lly) [reply src/core/core.js]: is setFilter or addFilter a better name?

Comment (Delerme) [reply src/core/core.js]: already changed to setFilter, unless my push got messed up

**Review by b1lly — APPROVED** [review]
LGTM as long as you manually tested both universal, vertical cases (search and auto complete).

**Pull request merged**
